holiday-calendar-base/src/test/java/com/github/davejoyce/calendar/SpecialAnniversaryTest.java

```diff
@@ -0,0 +1,111 @@
+
+package com.github.davejoyce.calendar;
+
+import lombok.extern.slf4j.Slf4j;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.time.LocalDate;
+import java.time.Month;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+import static org.testng.Assert.*;
+
+@Slf4j
+public class SpecialAnniversaryTest {
+
+    @Test
+    public void testConstructor_3Args() {
+        LocalDate jubileeDate = LocalDate.of(2022, Month.FEBRUARY, 6);
+        SpecialAnniversary holiday = new SpecialAnniversary("Platinum Jubilee of Elizabeth II",
+                "70th anniversary of the accession of Queen Elizabeth II", jubileeDate);
+        assertEquals(holiday.getAnniversaryDate(), jubileeDate);
+        assertEquals(holiday.getMonth(), Month.FEBRUARY);
+        assertEquals(holiday.getDayOfMonth(), 6);
+        assertFalse(holiday.isRollable());
+    }
+
+    @Test(dataProvider = "data")
+    public void testDateForYear(String name, LocalDate anniversaryDate, int yearToCalculate, LocalDate expected) {
+        SpecialAnniversary holiday = new SpecialAnniversary(name, "", anniversaryDate);
+        Optional<LocalDate> actual = holiday.dateForYear(yearToCalculate);
+        final String dateToLog;
+        if (null == expected) {
+            assertFalse(actual.isPresent());
+            dateToLog = "empty";
+        } else {
+            assertTrue(actual.isPresent());
+            assertEquals(actual.get(), expected);
+            assertEquals(holiday.getYear(), yearToCalculate);
+            dateToLog = actual.get().toString();
+        }
+        log.debug("{} {}: {}", name, yearToCalculate, dateToLog);
+    }
+
+    @Test
+    public void testEquals() {
+        LocalDate jubileeDate = LocalDate.of(2022, Month.FEBRUARY, 6);
+        LocalDate notJubileeDate = LocalDate.of(2021, Month.FEBRUARY, 6);
+        SpecialAnniversary holiday1 = new SpecialAnniversary("Platinum Jubilee of Elizabeth II", "70th anniversary of the accession of Queen Elizabeth II", jubileeDate);
+        SpecialAnniversary holiday2 = new SpecialAnniversary("Platinum Jubilee of Elizabeth II", "70th anniversary of the accession of Queen Elizabeth II", jubileeDate);
+        SpecialAnniversary holiday3 = new SpecialAnniversary("Platinum Jubilee of Elizabeth II", "70th anniversary of the accession of Queen Elizabeth II", notJubileeDate);
+        SpecialAnniversary holiday4 = new SpecialAnniversary("Platinum Jubilee of Elizabeth II", "70th anniversary", jubileeDate);
+        Object notAHoliday = new Object();
+
+        assertEquals(holiday1, holiday1);
+        assertEquals(holiday2, holiday1);
+        assertNotEquals(holiday1, null);
+        assertNotEquals(notAHoliday, holiday1);
+        assertNotEquals(holiday3, holiday1);
+        assertNotEquals(holiday4, holiday1);
+        assertFalse(holiday1.equals(null));
+        assertFalse(holiday1.equals(notAHoliday));
```

Comment:
While SonarLint complains that these could be simplified to use `assertNotEquals`, JaCoCo test coverage report shows that this line gets skipped in test of the  `equals` method if you don't use `assertFalse`:
```java
if (o == null || getClass() != o.getClass()) return false;
```

Reply:
Interesting.  That makes me wonder what `assertNotEquals` is actually doing then, since you'd think these should be equivalent.

Reply:
🤔 I too wonder... SonarLint is going to complain about this until we improve the assertions